src/target.rs

```diff
@@ -138,42 +137,136 @@ fn find_newest(path: &str) -> Result<File, Error> {
+                InputCase::New => time + Duration::from_secs(1),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, Copy)]
+    enum OutputCase {
+        None,
+        Missing,
+        Some,
+    }
+
+    fn all_test_cases() -> Vec<(InputCase, OutputCase)> {
+        let input_cases = [
+            InputCase::None,
+            InputCase::Old,
+            InputCase::Equal,
+            InputCase::New,
+        ];
+        let output_cases = [OutputCase::None, OutputCase::Missing, OutputCase::Some];
+
+        input_cases
+            .into_iter()
+            .flat_map(|ic| output_cases.map(|oc| (ic, oc)))
+            .collect()
+    }
+
+    fn setup_test_case(tempdir: &TempDir, ic: InputCase, oc: OutputCase) -> Vec<String> {
+        let now = SystemTime::now() - Duration::from_secs(60); // Back in time so avoid flaky tests
+        let mut args = vec![];
+
+        // Setup the output...
+        let output_path = tempdir
+            .path()
+            .join("output-file")
+            .to_str()
+            .unwrap()
+            .to_owned();
+
+        match oc {
+            OutputCase::None => (),
+            OutputCase::Missing => args.push(output_path.clone()),
+            OutputCase::Some => {
+                std::fs::File::create_new(&output_path)
+                    .unwrap()
+                    .set_modified(now)
+                    .unwrap();
+                args.push(output_path.clone());
+            }
+        }
+
+        // Setup the input...
+        args.push(":".into());
+        let input_path = tempdir
+            .path()
+            .join("input-file")
+            .to_str()
+            .unwrap()
+            .to_owned();
+
+        match ic {
+            InputCase::None => (),
+            ic @ _ => {
+                std::fs::File::create_new(&input_path)
+                    .unwrap()
+                    .set_modified(ic.permute_time(now))
+                    .unwrap();
+                args.push(input_path);
+            }
+        }
+
+        args.push("--".into());
+        args.push("touch".into());
+        args.push(output_path.into());
+
+        args
+    }
+
+    fn check_build(args: &Vec<String>, expected_needs_build: bool) {
+        let target = Target::parse(args.clone()).unwrap();
+        assert_eq!(target.needs_rebuild, expected_needs_build);
+        assert_eq!(target.should_run_command(), expected_needs_build);
+
+        let expected_code = 0;
+        let actual_code = target.run_command().unwrap();
+        assert_eq!(actual_code, expected_code);
+    }
+
+    #[tracing_test::traced_test]
     #[test]
-    fn test_target() {
-        let tmp_dir = tempdir().unwrap();
-
-        let src = tmp_dir.path().join("a.c");
-        let dest = tmp_dir.path().join("a.out");
-
-        std::fs::write(&src, "int main() { return 0; }").unwrap();
-
-        let target = Target::parse(
-            [
-                "target/debug/mk",
```

Comment:
This test _always_ runs the build because dropping `target/debug/mk` as `args[0]` is done in `main` and not in `parse`; `mk` is treated as a missing output.